pose_format/utils/holistic.py

```diff
@@ -52,11 +54,14 @@ def process_holistic(frames: list, fps: float, w: int, h: int, kinect=None, prog
         results = holistic.process(frame)
 
         body_data, body_confidence = body_points(results.pose_landmarks, w, h, 33)
-        face_data, face_confidence = component_points(results.face_landmarks, w, h, 468)
+        face_data, face_confidence = component_points(results.face_landmarks, w, h, 478 if holistic_config['refine_face_landmarks'] else 468)
         lh_data, lh_confidence = component_points(results.left_hand_landmarks, w, h, 21)
         rh_data, rh_confidence = component_points(results.right_hand_landmarks, w, h, 21)
+        # TODO: add pose_world_landmarks
+        # TODO: reduce the face landmarks?
 
         data = np.concatenate([body_data, face_data, lh_data, rh_data])
+        # TODO: information loss after concat?
```

Comment:
I am wondering whether this would lose the information that where a landmark comes from.

Reply:
Oh don't worry about that.

All of the landmarks are stored in one massive tensor `(frames, people, keypoints, dimensions)`, so they can all be manipulated together.
There are methods to "extract" a single keypoint by name and component. There is a way to separate it back, and as you see from the visualization, things are visualized differently (color) depending on what component they are from.

Having the large tensor instead of 4 increases read time almost by 4, and write time almost by 4 as well. This is because of the overhead for the "read" and "write" methods. Storing it this way is also more efficient for computation - min and max are all pooled, normalization, etc..